### ISSUE TITLE
Query with GROUP BY returns a single (too large) page 

### ISSUE THREAD

**Issue by avelanarius** [scylladb/scylla#7431]
```
cqlsh> CREATE KEYSPACE ks WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 1};
cqlsh> CREATE TABLE ks.t(pk int, PRIMARY KEY(pk));
cqlsh> INSERT INTO ks.t(pk) VALUES (1);
cqlsh> INSERT INTO ks.t(pk) VALUES (2);
cqlsh> INSERT INTO ks.t(pk) VALUES (3);
cqlsh> INSERT INTO ks.t(pk) VALUES (4);
cqlsh> INSERT INTO ks.t(pk) VALUES (5);
cqlsh> PAGING 1
Page size: 1
cqlsh> SELECT pk FROM ks.t GROUP BY pk;

 pk
----
  5
  1
  2
  4
  3

(5 rows)
```

Despite setting page size to 1, Scylla returns a single page of size 5.

Another example:
```
cqlsh> CREATE TABLE ks.t4(pk int, ck int, PRIMARY KEY(pk, ck));
cqlsh> INSERT INTO ks.t4(pk, ck) VALUES (1, 1);
cqlsh> INSERT INTO ks.t4(pk, ck) VALUES (1, 2);
cqlsh> INSERT INTO ks.t4(pk, ck) VALUES (2, 1);
cqlsh> INSERT INTO ks.t4(pk, ck) VALUES (3, 1);
cqlsh> INSERT INTO ks.t4(pk, ck) VALUES (4, 1);
cqlsh> INSERT INTO ks.t4(pk, ck) VALUES (4, 2);
cqlsh> PAGING 1
Page size: 1
cqlsh> SELECT pk FROM ks.t4 GROUP BY pk;

 pk
----
  1
  2
  4
  3

(4 rows)
cqlsh> SELECT pk, ck FROM ks.t4 GROUP BY pk, ck;

 pk | ck
----+----
  1 |  1
  1 |  2
  2 |  1
  4 |  1
  4 |  2
  3 |  1

(6 rows)
```

In those two SELECTs Scylla also returns a single (too large) page. On the other hand, Cassandra 3.11.8 correctly pages second and third query (Cassandra returns an error with regards to first query: "Invalid value for the paging state" when trying to fetch second page).

From what I could find, there is no information about this fact in documentation. If this is expected behaviour, maybe the documentation should warn about it.

**Comment by dekimir** [comment]
Likely related to #5362.  GROUPBY has trouble counting groups, because by the time it discovers one group has ended, it has already consumed a row from the next group, and there is no mechanism in the surrounding framework to defer the processing of the current row.

**Comment by dekimir** [comment]
> From what I could find, there is no information about this fact in documentation. If this is expected behaviour, maybe the documentation should warn about it.

I'd rather fix the bug than twist the documentation around it (and have to remember to untwist it in the future when the bug does get fixed).

**Comment by avikivity** [comment]
One ugly fix is to store the row that starts the next group in paging_state. When continuing the next page we retrieve the row, and start reading the next one.

**Comment by slivne** [comment]
On Sun, Oct 25, 2020 at 2:02 PM Avi Kivity <notifications@github.com> wrote:

> One ugly fix is to store the row that starts the next group in
> paging_state. When continuing the next page we retrieve the row, and start
> reading the next one.
>

That can cause you to skip some rows (and in some cases many rows) if
paging is done in parallel to writes

**Comment by avikivity** [comment]
> On Sun, Oct 25, 2020 at 2:02 PM Avi Kivity ***@***.***> wrote: One ugly fix is to store the row that starts the next group in paging_state. When continuing the next page we retrieve the row, and start reading the next one.
> That can cause you to skip some rows (and in some cases many rows) if paging is done in parallel to writes —


That's always true. In fact the best behavior would be to skip all rows that are written in parallel (to provide isolation).

**Comment by slivne** [comment]
In paging AFAIK we do not hold the next partition + next row

We hold the last partition + last row and start reading from that.

**Comment by avikivity** [comment]
Yes, but in this case we've already read an extra row. So we either have to drop and re-read it, or we have to store it somewhere.
